### DISCUSSION TITLE
Component Library Authors: How do you bundle Server components that contain client components?

### DISCUSSION THREAD

**Discussion by arkmech** [vercel/next.js#62231] [Help]
### Summary

I have solved the problem of adding `'use client';` when bundling my client components.

The problem: Bundling a server component that contains a client component.

Using tsup if that helps. My current belief is it is not fundamentally possible to make a component library that contains server components due to not having fine grained control of `use client` directive on bundled code.

### Additional information

Example: 
Part of my UI library contains a stereotypical Page Section.

`<PageSection />` has `title` and `description` props which are not reactive and should be server component.

It also contains a `<Button />` marked with `'use client';`.

When bundling, the consumer of the library needs to wrap the `<PageSection />` in `'use client';` due to the `<Button />` so we lose out on the benefits of server components OR I as the library author can add a `'use client';` myself, but also lose out on having a server components.

```ts
import { Button, ButtonProps } from '..';

type PageSectionProps = {
  title?: string;
  description?: string;
  buttonContent?: ButtonProps['children'];
};

export const PageSection = ({
  title,
  description,
  buttonContent,
}: PageSectionProps) => {
  return (
    <div>
      {(title || description) && (
        <div>
          {title}
          {description}
        </div>
      )}
      {buttonContent && <Button>{buttonContent}</Button>}
    </div>
  );
};
```

**Comment by arkmech** [comment]
Update, don’t think it’s possible

**Comment by MarciD** [comment]
Any updates on this? 😅

**Comment by Kevintjuhz** [comment]
Also interested in updates on this?

**Comment by arkmech** [comment]
No solution, but my example should show onClick handler as conditional check that causes the issue. Even if you just passing through to client component and not calling the function.

At this point, I think it’s just a large oversight with RSC
